### ISSUE TITLE
Fix `--fix` not reporting unfixable errors to stderr

### ISSUE THREAD

**Issue by CaitlinWeb** [stylelint/stylelint#6604]
### What steps are needed to reproduce the bug?

Honestly not sure if this was by design or a bug, but I would expect `stdin` to work similarly to linting files. When you lint files it fixes what it can in the files, and any remaining lint errors are output to `stdout`. Of course with `stdin` the changes are output to `stdout` as expected; it's the errors that aren't output at all, when I would expect them in `stderr`.

Use a CSS file that has lint errors that cannot be automatically fixed. Here's a contrived "test.css" that violates the unfixable `no-descending-specificity`, which is set to true by `stylelint-config-recommended`:

```css
.component {
  &.dark .header {
    color: white;
  }

  .header {
    color: black;
  }
}
```

Then in CLI: `stylelint --fix --stdin --stdin-filename "test.css" < "test.css"`

### What Stylelint configuration is needed to reproduce the bug?

Here's my current configuration in package.json:

```json
    "stylelint": {
        "extends": [
            "stylelint-config-standard",
            "stylelint-config-tailwindcss"
        ],
        "ignoreFiles": [
            "node_modules/**",
            "vendor/**"
        ],
        "rules": {
            "import-notation": ["string", { "severity": "warning", "disableFix": true }],
            "number-max-precision": 7,
            "selector-max-universal": 1,
            "value-keyword-case": ["lower", { "camelCaseSvgKeywords": true }]
        }
    },
```

### How did you run Stylelint?

CLI

### Which version of Stylelint are you using?

14.16.1

### What did you expect to happen?

`npx stylelint --fix --stdin --stdin-filename "test.css" < "test.css"` (output of file, output of errors):

<pre>.component {
  &.dark .header {
    color: white;
  }

 .header {
    color: black;
  }
}

test.css
 6:3  ✖  Expected selector ".component .header" to come before selector ".component.dark .header"  no-descending-specificity

1 problem (1 error, 0 warnings)</pre>

exit code is `2`

---

Or when outputting to files (easier way to see this), for example:

`npx stylelint --fix --stdin --stdin-filename "test.css" < "test.css" 1>changed-file.css 2>lint-errors.txt`

changed-file.css:
```css
.component {
  &.dark .header {
    color: white;
  }

  .header {
    color: black;
  }
}
```

lint-errors.txt:
```

test.css
 6:3  ✖  Expected selector ".component .header" to come before selector ".component.dark .header"  no-descending-specificity

1 problem (1 error, 0 warnings)
```

exit code is `2`

### What actually happened?

`npx stylelint --fix --stdin --stdin-filename "test.css" < "test.css"` (output of file only):

<pre>.component {
  &.dark .header {
    color: white;
  }

 .header {
    color: black;
  }
}</pre>

exit code is `2`

---

Or when outputting to files (easier way to see this), for example:

`npx stylelint --fix --stdin --stdin-filename "test.css" < "test.css" 1>changed-file.css 2>lint-errors.txt`

changed-file.css:
```css
.component {
  &.dark .header {
    color: white;
  }

  .header {
    color: black;
  }
}
```

lint-errors.txt (empty):
```
```

exit code is `2`

### Does the bug relate to non-standard syntax?

No

### Proposal to fix the bug

Take this with a grain of salt since I'm not familiar with all the code, but I think it would work to add the reported lint errors to `process.stderr.write` within this conditional for setting the exit code 🤷:
https://github.com/stylelint/stylelint/blob/f8988a10455be84b27c300631c902e289cfc730c/lib/cli.js#L543

**Comment by CaitlinWeb** [comment]
Some additional context with a workaround.

I'm setting up a pre-commit hook with Husky and `git-lint-staged`, which I have already done for standardjs which works perfectly in one line (applies fixes to staged files, fails the commit if there are lint errors, and shows the remaining errors).

I'm currently having to do this:

1) Apply automatic fixes to staged files and continue even if it exited with lint errors (`2`)
```shell
npx git-format-staged --formatter 'stylelint --fix --stdin --stdin-filename "{}" || if [ $? -eq 2 ]; then exit 0; fi' "*.css"
```

2) Lint for remaining errors, allowing any exit code so it appropriately blocks the commit if exit code is not `0`, and pipe them to `stderr` so they are displayed on exit:
```shell
npx git-format-staged --no-write --formatter 'stylelint --stdin --stdin-filename "{}" 1>&2' "*.css"
```

I couldn't figure out how to combine this into one line due to how `--fix` works, described above.

**Comment by jeddy3** [comment]
@CaitlinWeb Thank you for the very well-written issue.

> Honestly not sure if this was by design or a bug

I suspect we never considered this use case, so more likely a bug.

You're welcome to open a PoC pull request for your proposed change.

**Comment by haocheng6** [comment]
Hi @jeddy3, I am interested in working on this issue. Do you think the fixed code and unfixed errors/deprecation warnings should be output to stdout and stderr respectively in this case? Or should both of them be output to stdout? It looks like currently all linting errors of stylelint are output to stdout rather than stderr.

For context, different linting tools have different choices. ESLint outputs errors to stdout while deno lint outputs errors to stderr. Here is a related discussion: https://github.com/denoland/deno/issues/7304.

**Comment by jeddy3** [comment]
@haocheng6 Thank you for your interest.

> Do you think the fixed code and unfixed errors/deprecation warnings should be output to stdout and stderr respectively in this case?

Yes, I believe that's what's needed for the expected behaviour in the [original post](https://github.com/stylelint/stylelint/issues/6604#issue-1563366907), i.e.:

```
npx stylelint --fix --stdin --stdin-filename "test.css" < "test.css" 1>changed-file.css 2>lint-errors.txt
```

Sends the fixed CSS to `changed-file.css` and the lint problems to `lint-errors.txt`

Thanks also for the denolint link. The rationale that a linter's "expected result is nothing, and errors are always unexpected" makes sense to me.

Do we think this is a patch change or a breaking one for the `v16` branch? I can't imagine other tooling depending on the current behaviour of the CLI, but maybe @ybiquitous has thoughts on this.

**Comment by ybiquitous** [comment]
> Do we think this is a patch change or a breaking one for the `v16` branch? I can't imagine other tooling depending on the current behaviour of the CLI, but maybe @ybiquitous has thoughts on this.

Although I believe few tools rely on this behavior, the change could potentially break the compatibility. So personally, I recommend working on this issue on the `v16` branch.

> It should be fixed in the same version as #4989 and #3401.

Good point. Agree. 👍🏼

**Comment by ybiquitous** [comment]
Closed via #7119
